gufe/protocols/protocoldag.py

```diff
@@ -3,14 +3,93 @@
 
 import abc
 from typing import Iterable, List, Dict, Set, Optional, Union
+from os import PathLike
+import tempfile
 
 import networkx as nx
 
-from .protocolunit import ProtocolUnit
-from .results import ProtocolDAGResult, ProtocolDAGFailure
+from ..tokenization import GufeTokenizable
+from .protocolunit import ProtocolUnit, ProtocolUnitResult
 
 
-class ProtocolDAG:
+class ProtocolDAGResult(GufeTokenizable):
+    """Result for a single `ProtocolDAG` execution.
+
+    There may be many of these in a given `ResultStore` for a given `Transformation`.
+    Data elements from these objects are combined by `Protocol.gather` into a
+    `ProtocolResult`.
+
+    Attributes
+    ----------
+    name : str
+        Unique identifier for this `ProtocolDAGResult`.
+    graph : nx.DiGraph
+        The `ProtocolUnit`s, with dependencies set, as a networkx `DiGraph`.
+        Each `ProtocolUnit` features its `ProtocolUnitCompletion` as a `result` attribute.
+
+    """
+    name: Optional[str]
+    graph: nx.DiGraph
+
+    def __init__(self, *, name=None, graph):
+        self._name = name
+        self._graph = graph
+
+    def _defaults(self):
+        # not used by `ProtocolDAG`
+        return {}
+
+    def _to_dict(self):
+        return {'name': self.name,
+                'graph': self.graph}
+
+    @classmethod
+    def _from_dict(cls, dct: Dict):
+        return cls(**dct)
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def graph(self):
+        return self._graph
+
+    @property
+    def protocol_units(self):
+        return [pu for pu in self.graph.nodes]
+
+    @property
+    def protocol_unit_results(self):
+        return list(nx.get_node_attributes(self.graph, "result").values())
```

Comment:
Is this guaranteed to be the same order as above? (It ought to be)

Reply:
Now it should be, since we traverse the result graph in the same way. Will add test.